### ISSUE TITLE
openwrt transmission-daemon doesnt start

### ISSUE THREAD

**Issue by OwenG-AVISPL** [transmission/transmission#6387]
### What is the issue?

sorry i wanted to post on forums but unable to register, I'm not entirely sure this is where this belongs
OpenWrt 23.05.0 r23497-6637af95aa / LuCI openwrt-23.05 branch git-23.236.53405-fc638c8
install fresh the transmission-daemon (and luci app)
notice the daemon cant start at this point

initial error
> Wed Dec 6 13:27:29 2023 user.notice transmission: Starting with 505940000 virt mem
>Wed Dec 6 13:27:29 2023 user.err : jail: failed to load dependencies
>Wed Dec 6 13:27:29 2023 daemon.err transmission-daemon[25589]: jail: failed to load dependencies
>Wed Dec 6 13:27:29 2023 daemon.info procd: Instance transmission::instance1 s in a crash loop 9 crashes, 0 seconds since last crash

after which I tried transmission-daemon -f
which led me to install libmbedtls12 , transmission-cli & transmission-web
however the service is still failing to start

> root@GL-MT6000:~# transmission-daemon -f
> Error relocating /usr/bin/transmission-daemon: evthread_set_condition_callbacks: symbol not found
> Error relocating /usr/bin/transmission-daemon: evthread_set_id_callback: symbol not found
> Error relocating /usr/bin/transmission-daemon: evthread_set_lock_callbacks: symbol not found

some referenced post:  
[openwrt forums](https://forum.openwrt.org/t/transmission-daemon-errors/180068) 
[gl-inet forums](https://forum.gl-inet.com/t/gl-mt6000-flint2-transmission-daemon-error/35898) 

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.3-4

**Comment by dmrlx** [comment]
Absolutely the same for me. 

```
root@GL-MT6000:/tmp# service transmission status
running
```
but
```
root@GL-MT6000:/tmp# transmission-daemon
Error relocating /usr/bin/transmission-daemon: evthread_set_condition_callbacks: symbol not found
Error relocating /usr/bin/transmission-daemon: evthread_set_id_callback: symbol not found
Error relocating /usr/bin/transmission-daemon: evthread_set_lock_callbacks: symbol not found
```

**Comment by dmrlx** [comment]
I found following issue https://github.com/transmission/transmission/issues/3948, but author there says that it was necessary to add dependencies. I've tried, but it doesn't help.

**Comment by dmrlx** [comment]
It looks like I found solution. Necessary calls (evthread_set_condition_callbacks, evthread_set_id_callback and evthread_set_lock_callbacks) are in another lib, "libevent2-extra7" and you need to install it and change one link on your system.

What I had originally:

```
root@GL-MT6000:~# ldd /usr/bin/transmission-daemon
	/lib/ld-musl-aarch64.so.1 (0x7f92283000)
	libatomic.so.1 => /lib/libatomic.so.1 (0x7f92261000)
	libevent-2.1.so.7 => /usr/lib/libevent-2.1.so.7 (0x7f92210000)
	libdeflate.so.0 => /usr/lib/libdeflate.so.0 (0x7f921ef000)
	libmbedcrypto.so.7 => /usr/lib/libmbedcrypto.so.7 (0x7f9217c000)
	libcurl.so.4 => /usr/lib/libcurl.so.4 (0x7f920f9000)
	libpsl.so.5 => /usr/lib/libpsl.so.5 (0x7f920c8000)
	libnatpmp.so.1 => /usr/lib/libnatpmp.so.1 (0x7f920a7000)
	libminiupnpc.so.17 => /usr/lib/libminiupnpc.so.17 (0x7f92086000)
	libdht.so => /usr/lib/libdht.so (0x7f92065000)
	libutp.so => /usr/lib/libutp.so (0x7f92040000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x7f91e6c000)
	libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x7f91e3b000)
	libc.so => /lib/ld-musl-aarch64.so.1 (0x7f92283000)
	libssl.so.3 => /usr/lib/libssl.so.3 (0x7f91d77000)
	libcrypto.so.3 => /usr/lib/libcrypto.so.3 (0x7f91912000)
	libidn2.so.0 => /usr/lib/libidn2.so.0 (0x7f918d1000)
	libunistring.so.5 => /usr/lib/libunistring.so.5 (0x7f91710000)
Error relocating /usr/bin/transmission-daemon: evthread_set_condition_callbacks: symbol not found
Error relocating /usr/bin/transmission-daemon: evthread_set_id_callback: symbol not found
Error relocating /usr/bin/transmission-daemon: evthread_set_lock_callbacks: symbol not found
```
Then I performed described actions
```
# opkg install libevent2-extra7
# rm -rf /usr/lib/libevent-2.1.so.7
# ln -s /usr/lib/libevent_extra-2.1.so.7.0.1 /usr/lib/libevent-2.1.so.7
```
And after that:
```
root@GL-MT6000:~# ldd /usr/bin/transmission-daemon
	/lib/ld-musl-aarch64.so.1 (0x7f92866000)
	libatomic.so.1 => /lib/libatomic.so.1 (0x7f92844000)
	libevent-2.1.so.7 => /usr/lib/libevent-2.1.so.7 (0x7f92813000)
	libdeflate.so.0 => /usr/lib/libdeflate.so.0 (0x7f927f2000)
	libmbedcrypto.so.7 => /usr/lib/libmbedcrypto.so.7 (0x7f9277f000)
	libcurl.so.4 => /usr/lib/libcurl.so.4 (0x7f926fc000)
	libpsl.so.5 => /usr/lib/libpsl.so.5 (0x7f926cb000)
	libnatpmp.so.1 => /usr/lib/libnatpmp.so.1 (0x7f926aa000)
	libminiupnpc.so.17 => /usr/lib/libminiupnpc.so.17 (0x7f92689000)
	libdht.so => /usr/lib/libdht.so (0x7f92668000)
	libutp.so => /usr/lib/libutp.so (0x7f92643000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x7f9246f000)
	libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x7f9243e000)
	libc.so => /lib/ld-musl-aarch64.so.1 (0x7f92866000)
	libevent_core-2.1.so.7 => /usr/lib/libevent_core-2.1.so.7 (0x7f923fd000)
	libssl.so.3 => /usr/lib/libssl.so.3 (0x7f92339000)
	libcrypto.so.3 => /usr/lib/libcrypto.so.3 (0x7f91ed4000)
	libidn2.so.0 => /usr/lib/libidn2.so.0 (0x7f91e93000)
	libunistring.so.5 => /usr/lib/libunistring.so.5 (0x7f91cd2000)
```

**Comment by tearfur** [comment]
Yes, the cause is just missing libraries on your system.

**Comment by dmrlx** [comment]
In my opinion, the fact that we need to substitute the library speaks to a problem in transmission dependencies.

**Comment by tearfur** [comment]
It looks like the problem is with packaging, instead of our program code though.

**Comment by dmrlx** [comment]
Uh, Ok, got it.

**Comment by dmrlx** [comment]
Uh, Ok, got it.

**Comment by OwenG-AVISPL** [comment]
Thank you guys very much
however i still seem to have issues (cant connect to daemon)
i see this error in logs

> [2023-12-29 07:22:08.225] ERR crypto-utils-mbedtls.cc:232 MbedTLS error: CTR_DRBG - The requested random buffer length is too big (-54) (crypto-utils-mbedtls.cc:232)

googling lead me to this post, which is closed but i dont see any suggestions on what i can do to resolve it, and its from 2022
 https://github.com/transmission/transmission/issues/5795

**Comment by mikedld** [comment]
The fix (#6379) was recently pushed to `main` branch and will be included in either 4.0.6 (if we decide to make yet another 4.0.x patch release) or 4.1. You can grab the patch if you're building it yourself and see if it makes things better.

**Comment by jjm2473** [comment]
This is a pitfall of openwrt. If you compile the openwrt firmware yourself with libevent2 built in, but do not compile libevent2-pthreads, then the pthread-related functions will be turned off in libevent2.

https://github.com/openwrt/openwrt/blob/92d4f59709eac7422391187edbd85f2a14d5589b/package/libs/libevent2/Makefile#L125-L126

The Transmission you installed from the OpenWrt feed depends on libevent2, which includes pthread support (because libevent2-pthreads is compiled with Transmission). However, the libevent2 in your firmware does not support pthread, and opkg will not install the libevent2 from the feed (because they are the same version). Therefore, the installed Transmission will not run.
